fcrepo4/fcrepo4
# moved ocfl index into db

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3309 & https://jira.lyrasis.org/browse/FCREPO-3292

# What does this Pull Request do?

Moves the index that maps Fedora IDs to OCFL ids into the database. Additionally, the rebuild logic was adjusted so that both the containment index and the ocfl index are rebuilt when they are out of sync with the OCFL repository.

# How should this be tested?

1. Startup a DB and Fedora using the following instructions.
1. Add some resources
1. Inspect the index in the db
1. Stop Fedora and restart it again using a different db configuration
1. Inspect the index in the new db and see that the index is populated

## MariaDB

Start the DB:

```
docker run --name f6-mariadb -e MYSQL_ROOT_PASSWORD=root-pw -e MYSQL_DATABASE=fcrepo -e MYSQL_USER=fcrepo-user -e MYSQL_PASSWORD=fcrepo-pw -p 3306:3306 -d mariadb:10.5.3
```

Access the CLI:

```
docker exec -it f6-mariadb mysql -ufcrepo-user -pfcrepo-pw -Dfcrepo
```

Start F6:

```
mvn -Dfcrepo.db.url=jdbc:mariadb://localhost:3306/fcrepo -Dfcrepo.db.user=fcrepo-user -Dfcrepo.db.password=fcrepo-pw clean jetty:run
```

Stop DB & cleanup:

```
docker stop f6-mariadb
docker rm f6-mariadb
```

## MySQL

Start the DB:

```
docker run --name f6-mysql -e MYSQL_ROOT_PASSWORD=root-pw -e MYSQL_DATABASE=fcrepo -e MYSQL_USER=fcrepo-user -e MYSQL_PASSWORD=fcrepo-pw -p 3306:3306 -d mysql:8.0
```

Access the CLI:

```
docker exec -it f6-mysql mysql -ufcrepo-user -pfcrepo-pw -Dfcrepo
```

Start F6:

```
mvn -Dfcrepo.db.url=jdbc:mysql://localhost:3306/fcrepo -Dfcrepo.db.user=fcrepo-user -Dfcrepo.db.password=fcrepo-pw clean jetty:run
```

Stop DB & cleanup:

```
docker stop f6-mysql
docker rm f6-mysql
```

## PostgreSQL

Start the DB:

```
docker run --name f6-postgres -e POSTGRES_USER=fcrepo-user -e POSTGRES_PASSWORD=fcrepo-pw -e POSTGRES_DB=fcrepo -p 5432:5432 -d postgres:12.3
```

Access the CLI:

```
docker exec -it f6-postgres psql -U fcrepo-user -d fcrepo
```

Start F6:

```
mvn -Dfcrepo.db.url=jdbc:postgresql://localhost:5432/fcrepo -Dfcrepo.db.user=fcrepo-user -Dfcrepo.db.password=fcrepo-pw clean jetty:run
```

Stop DB & cleanup:

```
docker stop f6-postgres
docker rm f6-postgres
```

# Interested parties

@fcrepo4/committers
